File: 437_PathSumIII.cpp
//69.41%
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int pathSum(TreeNode* root, int sum) {
        if(!root)
			return 0;
		return pathSumHelper(root, sum) + pathSum(root -> left, sum) + pathSum(root -> right, sum);	
    }
private:
	int pathSumHelper(TreeNode *root, int sum){
		if(!root)
			return 0;
		int count(0);
		if(sum == root -> val)
			count++;
		count += pathSumHelper(root -> left, sum - root -> val);
		count += pathSumHelper(root -> right, sum - root -> val);
		return count;
	}
};

File: 437_PathSumIII.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
# 25.84%
class Solution(object):
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: int
        """
        if not root:
            return 0
        return self.pathSumHelper(root, sum) + self.pathSum(root.left, sum) + self.pathSum(root.right, sum)
    def pathSumHelper(self, root, sum):
        if not root:
            return 0
        count = 0
        if root.val == sum:
            count += 1
        return count + self.pathSumHelper(root.left, sum - root.val) + self.pathSumHelper(root.right, sum - root.val)